Content.Shared/Movement/Pulling/Systems/PullingSystem.cs

```diff
@@ -892,41 +919,31 @@ public void StopAllPulls(EntityUid uid) // Goobstation
     /// <param name="escapeAttemptModifier">if anything what to modify the escape chance by</param>
     /// <exception cref="ArgumentOutOfRangeException"></exception>
     /// <returns></returns>
-    public bool TryGrab(Entity<PullableComponent?> pullable, Entity<PullerComponent?> puller, bool ignoreCombatMode = false, GrabStage? grabStageOverride = null, float escapeAttemptModifier = 1f)
+    private bool TryGrab(Entity<PullableComponent?> pullable, Entity<PullerComponent?> puller, bool ignoreCombatMode = false, GrabStage? grabStageOverride = null, float escapeAttemptModifier = 1f)
     {
-        if (!Resolve(pullable.Owner, ref pullable.Comp))
-            return false;
-
-        if (!Resolve(puller.Owner, ref puller.Comp))
+        if (!Resolve(pullable.Owner, ref pullable.Comp)
+            || !Resolve(puller.Owner, ref puller.Comp)
+            || HasComp<PacifiedComponent>(puller)
+            || !HasComp<MobStateComponent>(pullable)
+            || pullable.Comp.Puller != puller
+            || puller.Comp.Pulling != pullable
+            || !TryComp<MeleeWeaponComponent>(puller, out var meleeWeaponComponent))
             return false;
 
         // prevent you from grabbing someone else while being grabbed
         if (TryComp<PullableComponent>(puller, out var pullerAsPullable) && pullerAsPullable.Puller != null)
             return false;
 
-        if (HasComp<PacifiedComponent>(puller))
-            return false;
-
-        if (pullable.Comp.Puller != puller ||
-            puller.Comp.Pulling != pullable)
-            return false;
-
-        if (puller.Comp.NextStageChange > _timing.CurTime)
-            return true;
-
-        // You can't choke crates
-        if (!HasComp<MobStateComponent>(pullable))
-            return false;
-
-        // Delay to avoid spamming
-        puller.Comp.NextStageChange = _timing.CurTime + puller.Comp.StageChangeCooldown;
-        Dirty(puller);
-
         // Don't grab without grab intent
         if (!ignoreCombatMode)
             if (!_combatMode.IsInCombatMode(puller))
                 return false;
 
+        if (_timing.CurTime < meleeWeaponComponent.NextAttack)
+            return true;
+
+        meleeWeaponComponent.NextAttack += puller.Comp.StageChangeCooldown;
```

Comment:
```suggestion
        meleeWeaponComponent.NextAttack = puller.Comp.StageChangeCooldown + (meleeWeaponComponent.NextAttack > _timing.CurTime ? meleeWeaponComponent.NextAttack : _timing.CurTime);
```
If NextAttack is zero or something, this allows you to grab without cooldown.